Sources/StreamChatUI/CommonViews/Attachments/AttachmentCollectionViewCell.swift

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import StreamChat
+import UIKit
+
+public protocol AttachmentView: _View {
+    var content: AttachmentPreview? { get set }
+    var discardButtonHandler: (() -> Void)? { get set }
+}
+
+public typealias AttachmentCollectionViewCell = _AttachmentCollectionViewCell<NoExtraData>
+
+open class _AttachmentCollectionViewCell<ExtraData: ExtraDataTypes>: _CollectionViewCell, ComponentsProvider {
+    class var reuseId: String { String(describing: self) }
+
+    open var content: AttachmentPreview? {
+        didSet {
+            updateContentIfNeeded()
+        }
+    }
+    
+    public var discardAction: (() -> Void)?
+    
+    public private(set) var attachmentView: AttachmentView?
+    
+    open func replaceAttachmentView(with view: AttachmentView?) {
+        attachmentView?.removeFromSuperview()
+        attachmentView = view
+        
+        guard let view = view else { return }
+        embed(view)
+    }
+    
+    override open func prepareForReuse() {
+        super.prepareForReuse()
+
+        // Remove the attachment view from the view hiearchy, but keep it in memory.
+        // There is a high chance attachment view will be reused and embeded again.
+        attachmentView?.removeFromSuperview()
+    }
+    
+    open func attachmentView(for attachmentType: AttachmentType) -> AttachmentView? {
+        let type: AttachmentView.Type
+        
+        switch attachmentType {
+        case .image:
+            type = components.messageComposer.imageAttachmentView
+
+        case .file:
+            type = components.messageComposer.fileAttachmentView
+
+        default:
+            log.error("Unsuported attachment type: \(attachmentType)")
+            return nil
+        }
+        
+        // Cell reuse optimalisation
+        // Attachment view is already the right type. Don't create it again, just prepare it for reuse.
+        if attachmentView?.isKind(of: type) == true {
+            attachmentView?.content = nil
+            attachmentView?.discardButtonHandler = nil
+            return attachmentView
+        }
+        
+        let result: AttachmentView = type.init()
+        result.translatesAutoresizingMaskIntoConstraints = false
+        return result
+    }
```

Comment:
Wouldn't it be better if this happened on the controller level?

Reply:
Now it's happening on the attachment level, as per discussion